pkg/api/customization/cluster/validator.go

```diff
@@ -118,6 +125,58 @@ func (v *Validator) validateEnforcement(request *types.APIContext, data map[stri
 	return nil
 }
 
+// TODO: test validator
+// prevents downgrades, no-ops, and upgrading before versions have been set
+func (v *Validator) validateK3sVersionUpgrade(request *types.APIContext, spec *v3.ClusterSpec) error {
+	upgradeNotReadyErr := httperror.NewAPIError(httperror.Conflict, "k3s version upgrade is not ready, try again later")
+
+	if request.Method == http.MethodPost {
+		return nil
+	}
+
+	if spec.K3sConfig == nil {
+		// only applies to k3s clusters
+		return nil
+	}
+
+	// must wait for original spec version to be set
```

Comment:
this comment is weird. you arent waiting on anything; you're just ensuring the user didnt send a ""

Reply:
and more important, returning `upgradeNotReadyErr` is wrong in this case